### PR TITLE
Fix crash while quitting in AutoFileMatcher, #5056

### LINKED ISSUE #5056: Crash while quitting in AutoFileMatcher
**System and IINA version:**

- macOS 13.6.3
- IINA `develop` with some changes that were being tested

**Expected behavior:**
IINA does not crash.

**Actual behavior:**
IINA crashed while quitting when a background thread was still finding and adding files to the playlist.

As can be seen in the portion of the crash report below, the mpv client was being destroyed at the same time the background thread was running `AutoFileMatcher.addFilesToPlaylist`.

<details>
<summary><b>Crash report:</b></summary>

```
Thread 0::  Dispatch queue: com.apple.main-thread
0   libsystem_kernel.dylib        	       0x195ec6bbc madvise + 8
1   libsystem_malloc.dylib        	       0x195d2f064 mvm_madvise_free + 108
2   libsystem_malloc.dylib        	       0x195d2d720 free_medium + 2444
3   libmpv.2.dylib                	       0x1059a9aa0 ta_free_children + 88
4   libmpv.2.dylib                	       0x1059a990c ta_free + 104
5   libmpv.2.dylib                	       0x10594bc54 mp_destroy + 240
6   libmpv.2.dylib                	       0x10592ed80 mp_destroy_client + 940
7   IINA                          	       0x1012cb7dc MPVController.mpvUninitRendering() + 232 (MPVController.swift:692)
8   IINA                          	       0x1011843a4 closure #1 in VideoView.uninit() + 480 (VideoView.swift:96)
9   IINA                          	       0x1011843e0 partial apply for closure #1 in VideoView.uninit() + 24
10  IINA                          	       0x100fe06ac closure #1 in Atomic.withLock<A>(_:) + 136 (Atomic.swift:32)
11  IINA                          	       0x100fe0730 partial apply for closure #1 in Atomic.withLock<A>(_:) + 44
12  IINA                          	       0x10117873c OSUnfairLockImpl.withLock<A>(_:) + 104 (Lock.swift:70)
13  IINA                          	       0x101178894 protocol witness for LockImpl.withLock<A>(_:) in conformance OSUnfairLockImpl + 32
14  IINA                          	       0x1011784a4 Lock.withLock<A>(_:) + 144 (Lock.swift:42)
15  IINA                          	       0x100fe05bc Atomic.withLock<A>(_:) + 168 (Atomic.swift:31)
16  IINA                          	       0x10118418c VideoView.uninit() + 132 (VideoView.swift:91)
17  IINA                          	       0x10123447c PlayerCore.uninitVideo() + 484 (PlayerCore.swift:519)
18  IINA                          	       0x10123491c PlayerCore.mpvHasShutdown(isMPVInitiated:) + 604 (PlayerCore.swift:552)
19  IINA                          	       0x1012d7e64 closure #1 in MPVController.handleEvent(_:) + 104 (MPVController.swift:1051)
20  IINA                          	       0x100ddd814 thunk for @escaping @callee_guaranteed () -> () + 48
21  libdispatch.dylib             	       0x195d52874 _dispatch_call_block_and_release + 32
22  libdispatch.dylib             	       0x195d54400 _dispatch_client_callout + 20
23  libdispatch.dylib             	       0x195d62bf8 _dispatch_main_queue_drain + 928
24  libdispatch.dylib             	       0x195d62848 _dispatch_main_queue_callback_4CF + 44
25  CoreFoundation                	       0x196023bc4 __CFRUNLOOP_IS_SERVICING_THE_MAIN_DISPATCH_QUEUE__ + 16
26  CoreFoundation                	       0x195fe1344 __CFRunLoopRun + 1992
27  CoreFoundation                	       0x195fe0428 CFRunLoopRunSpecific + 612
28  HIToolbox                     	       0x19f834df0 RunCurrentEventLoopInMode + 292
29  HIToolbox                     	       0x19f834c2c ReceiveNextEventCommon + 648
30  HIToolbox                     	       0x19f834984 _BlockUntilNextEventMatchingListInModeWithFilter + 76
31  AppKit                        	       0x199208908 _DPSNextEvent + 636
32  AppKit                        	       0x199207aa4 -[NSApplication(NSEvent) _nextEventMatchingEventMask:untilDate:inMode:dequeue:] + 716
33  AppKit                        	       0x19942917c -[NSApplication _shouldTerminate] + 700
34  AppKit                        	       0x19953f894 -[NSApplication(NSAppleEventHandling) _handleAEQuit] + 268
35  AppKit                        	       0x19920d818 -[NSApplication(NSAppleEventHandling) _handleCoreEvent:withReplyEvent:] + 672
36  Foundation                    	       0x196f30b14 -[NSAppleEventManager dispatchRawAppleEvent:withRawReply:handlerRefCon:] + 316
37  Foundation                    	       0x196f30974 _NSAppleEventManagerGenericHandler + 80
38  AE                            	       0x19c30f55c 0x19c303000 + 50524
39  AE                            	       0x19c30ee88 0x19c303000 + 48776
40  AE                            	       0x19c307f5c aeProcessAppleEvent + 488
41  HIToolbox                     	       0x19f844cf4 AEProcessAppleEvent + 68
42  AppKit                        	       0x199208c0c _DPSNextEvent + 1408
43  AppKit                        	       0x199207aa4 -[NSApplication(NSEvent) _nextEventMatchingEventMask:untilDate:inMode:dequeue:] + 716
44  AppKit                        	       0x1991fbf08 -[NSApplication run] + 464
45  AppKit                        	       0x1991d3358 NSApplicationMain + 880
46  IINA                          	       0x100ffa0ac main + 12 (AppDelegate.swift:24)
47  dyld                          	       0x195babf28 start + 2236
…
Thread 2 Crashed::  Dispatch queue: IINAPlayerCoreTask
0   libmpv.2.dylib                	       0x105930754 mpv_get_property + 16
1   IINA                          	       0x1012d0c28 MPVController.getInt(_:) + 572 (MPVController.swift:833)
2   IINA                          	       0x100ec850c AutoFileMatcher.addFilesToPlaylist() + 2956 (AutoFileMatcher.swift:129)
3   IINA                          	       0x100ecfaa4 AutoFileMatcher.startMatching() + 2020 (AutoFileMatcher.swift:348)
4   IINA                          	       0x101256664 PlayerCore.autoLoadFilesInCurrentFolder(ticket:) + 92 (PlayerCore.swift:2018)
5   IINA                          	       0x10124fc18 closure #3 in PlayerCore.fileStarted(path:) + 416 (PlayerCore.swift:1739)
6   IINA                          	       0x100f45c18 thunk for @escaping @callee_guaranteed @Sendable () -> () + 48
7   libdispatch.dylib             	       0x195d52874 _dispatch_call_block_and_release + 32
8   libdispatch.dylib             	       0x195d54400 _dispatch_client_callout + 20
9   libdispatch.dylib             	       0x195d5ba88 _dispatch_lane_serial_drain + 668
10  libdispatch.dylib             	       0x195d5c5f8 _dispatch_lane_invoke + 384
11  libdispatch.dylib             	       0x195d67244 _dispatch_workloop_worker_thread + 648
12  libsystem_pthread.dylib       	       0x195f00074 _pthread_wqthread + 288
13  libsystem_pthread.dylib       	       0x195efed94 start_wqthread + 8
```
</details>

<br>

**Steps to reproduce:**
This was reproduced by running a script that starts and quits IINA over and over again to test for proper shutdown coordination.

- [x] MPV does not have this problem.

**How often does this happen?**
Rarely during my testing. Likely a test can be made that would make this more likely to occur.

### PR DESCRIPTION
**Pull request by low-batt**
This commit will:
- Correct the `active` computed property in `PlayerState` to not consider the stopping state as active
- Change the `PlaybackInfo.state` computed property to prevent additional inappropriate state transitions
- Change `PlayerCore.sendOSD` to not show messages unless the player is active to prevent an unreported crash
- Change `MPVController.readEvents` to stop once a `MPV_EVENT_SHUTDOWN` has been processed
- Change the `fileLoaded` and `fileStarted` `PlayerCore` methods to do nothing if the player state is not active
- Change the logger subsystem for `AutoFileMatcher` to include the player identifier
- Change the names of the `PlayerCore` dispatch queues to include the player identifier
- Move `TicketExpiredError` from `AutoFileMatcher` to `PlayerCore`
- Add a `checkTicket` method to `PlayerCore`
- Change more` AutoFileMatcher` methods to check the ticket and throw errors
- Change `AutoFileMatcher.startMatching` to rethrow errors
- Change `PlayerCore.fileStarted` to check the background queue ticket and catch errors
- Change `PlayerCore.fileStarted` to have the background task queue a main thread task that will continue the process of stopping the player if it was waiting for the background task to finish
- Change `PlayerCore.findIdlePlayerCore` to not choose an idle core if it's background task is still running

These changes cause the player to not send a stop command to mpv until the background task has stopped. A mpv core will stop on its own when the end of a video is reached. For that reason these changes also cause the player to not send a quit command to mpv until the background task has stopped. These changes correct crashes that happens when the background task is still running when the mpv core is quitting. This can occur for example, when a user starts playing the wrong video and immediately quits IINA before the background task has a chance to finish.

- [x] I have read [CONTRIBUTING.md](https://github.com/iina/iina/blob/develop/CONTRIBUTING.md)
- [x] This implements/fixes issue #5056.

---

**Description:**

### CODE REVIEW

File: iina/PlayerState.swift [hunk 1]
```diff
@@ -58,7 +58,7 @@ enum PlayerState: Int {
   ///
   /// These are the states in which the player normally interacts with the mpv core. The mpv core **must not** be accessed when the
   /// player is in the `shuttingDown` or `shutDown` states. Accessing the core in these states can trigger a crash.
-  @inlinable var active: Bool { self.rawValue <= PlayerState.stopping.rawValue }
+  @inlinable var active: Bool { self.rawValue < PlayerState.stopping.rawValue }
```
Comment (uiryuu) [inline iina/PlayerState.swift]: Please revert this line since it's already fixed in dev.

File: iina/MPVController.swift [hunk 2]
```diff
@@ -1024,18 +1028,23 @@ class MPVController: NSObject {
   private func readEvents() {
     queue.async {
       while ((self.mpv) != nil) {
-        let event = mpv_wait_event(self.mpv, 0)
+        let event = mpv_wait_event(self.mpv, 0)!
```
Comment (uiryuu) [inline iina/MPVController.swift:1006]: Is it safe to force unwarp here?

File: iina/PlayerCore.swift [hunk 3]
```diff
@@ -1716,30 +1780,58 @@ class PlayerCore: NSObject {
...
-        self.autoLoadFilesInCurrentFolder(ticket: currentTicket)
-      }
-      // auto load matched subtitles
-      if let matchedSubs = self.info.getMatchedSubs(path) {
-        self.log("Found \(matchedSubs.count) subs for current file")
-        for sub in matchedSubs {
-          guard currentTicket == self.backgroundQueueTicket else { return }
-          self.loadExternalSubFile(sub)
+    backgroundTaskInUse = true
+    backgroundQueue.async { [self] in
```
Comment (uiryuu) [inline iina/PlayerCore.swift:1782]: Consider to use `[unowned self]` here?

File: iina/PlayerCore.swift [hunk 4]
```diff
@@ -1716,30 +1780,58 @@ class PlayerCore: NSObject {
...
+          setTrack(1, forType: .sub)
+        }
+        autoSearchOnlineSub()
+      } catch TicketExpiredError.ticketExpired {
+        log("Background task stopping due to ticket expiration")
+      } catch let err {
+        log("Background task stopping due to error \(err.localizedDescription)", level: .error)
+      }
+      // This code must be queued to the main thread to avoid thread data races.
+      DispatchQueue.main.async { [self] in
```
Comment (uiryuu) [inline iina/PlayerCore.swift:1807]: Ditto

**Review by uiryuu — CHANGES_REQUESTED** [review]

File: iina/PlayerState.swift (hunk 1 above)
Comment (low-batt) [reply iina/PlayerState.swift]: I will rebase with dev.

File: iina/MPVController.swift (hunk 2 above)
Comment (low-batt) [reply iina/MPVController.swift:1006]: The change here is that the code was treating the return as option in `readEvents`, but then passed it to `handleEvent` which declared the type as `UnsafePointer<mpv_event>!`. So `readEvents` went to the trouble of treating it as optional only for it to be forced when `handleEvent` was called. Thus the forcing just got moved to an earlier time. I don't think there is a change as far as risking a crash due to force unwrapping.

If this is ever `nil` then mpv is seriously broken as the doc for `mpv_wait_event` says:
> The return value is never NULL.

File: iina/PlayerCore.swift (hunk 3 above)
Comment (low-batt) [reply iina/PlayerCore.swift:1782]: We never free players until shutdown so I don't see an advantage to the background task not having a strong reference.

File: iina/PlayerState.swift (hunk 1 above)
Comment (low-batt) [reply iina/PlayerState.swift]: Rebased with `develop` so this change is no longer in the commit. I amended the commit comment and removed the statement about changing `active`.

**Pull request merged**